### PR TITLE
Move observation to ngrx store

### PR DESCRIPTION
**Pull request by smadbe**
## Description

Move observation (and router to support observation) to ngrx store 

## Test cases

- [x] Case 1:
  1. Given I am the usual user
  2. I open the dev tools
  3. When I go to [the home page](https://dev.algorea.org/branch/rootstore/en/a/home;pa=0)
  4. And I click on "ALGOREA ADVENTURE "
  5. Then if I move back and forth in the store history, I navigate from one page to the other 

- [x] Case 2:
  1. Given I am the usual user
  2. I open the dev tools
  3. When I [start observation through url](https://dev.algorea.org/branch/rootstore/en/a/home;pa=0?watchedGroupId=672913018859223173&watchUser=0)
  4. I see the group name has been fetched and observation has started with the right name in the bar
  5. And I close the observation mode via the top bar
  6. And I am not in observation mode anymore

- [x] Case 3:
  1. Given I am the usual user
  3. When I [start observing a group that does not exist](https://dev.algorea.org/branch/rootstore/en/a/home;pa=0?watchedGroupId=404&watchUser=0)
  4. I see the error message
  7. and closer the modal continues without observation

- [x] Case 4:
  1. Given I am the usual user
  3. When I [start observing a group that I can see but not watch](https://dev.algorea.org/branch/rootstore/en/a/home;pa=0?watchedGroupId=52767158366271444&watchUser=0)
  4. I see the error message
  5. and closer the modal continues without observation

- [x] Case 5:
  1. Given I am the usual user
  2. And I open dev tool
  3. When I [go the page of a group I can watch](https://dev.algorea.org/branch/rootstore/en/groups/by-id/672913018859223173;p=52767158366271444)
  4. and I click on the eye in the top bar
  5. I see I am in observation mode
  8. I see in the network request that no request to get the group name has been done
  9. I go to "content" in the left menu
  10. I close the observation mode through the left menu
  11. And it leaves observation

- [x] Case 6:
  1. Given I am the usual user
  3. When I [go the page of a user I can watch](https://dev.algorea.org/branch/rootstore/en/groups/users/752024252804317630)
  4. and I click on the eye in the header
  5. I see I am in observation mode
  6. And I reclick on the eye
  7. And I leave the observation mode

### CODE REVIEW

File: src/app/items/item-by-id.component.html [hunk 1]
```diff
@@ -84,7 +84,7 @@
           ></alg-item-task-edit>
 
           <ng-container *ngIf="currentTab && [ 'alg-log', 'alg-chapter-progress'].includes(currentTab.tag)">
-            <ng-container *ngIf="(!isWatching && (itemData.item.permissions | allowsViewingContent)) || (itemData.item.permissions | allowsWatchingResults) ; else unallowed">
+            <ng-container *ngIf="(!isObserving$ && (itemData.item.permissions | allowsViewingContent)) || (itemData.item.permissions | allowsWatchingResults) ; else unallowed">
```
Comment (Iloveall) [inline src/app/items/item-by-id.component.html:87]: *ngrxLet="isObserving$ as isObserving": is not using isObserving here

File: src/app/items/item-by-id.component.html [hunk 2]
```diff
@@ -102,23 +102,23 @@
               <ng-container *ngIf="currentTab?.tag === 'alg-chapter-progress'">
                 <alg-chapter-group-progress
                   class="alg-flex-1"
-                  *ngIf="isWatching"
+                  *ngIf="isObserving$"
```
Comment (Iloveall) [inline src/app/items/item-by-id.component.html:105]: Same here

File: src/app/items/item-by-id.component.html [hunk 3]
```diff
@@ -102,23 +102,23 @@
...
                   [itemData]="itemData"
                 ></alg-chapter-group-progress>
 
                 <alg-chapter-user-progress
                   class="alg-flex-1"
-                  *ngIf="!isWatching"
+                  *ngIf="!isObserving$"
```
Comment (Iloveall) [inline src/app/items/item-by-id.component.html:111]: Same here

File: src/app/items/item-by-id.component.html [hunk 4]
```diff
@@ -102,23 +102,23 @@
...
                   [itemData]="itemData"
                 ></alg-chapter-user-progress>
               </ng-container>
 
             </ng-container>
             <ng-template #unallowed>
-              <p class="not-allow-caption" *ngIf="isWatching" i18n>
+              <p class="not-allow-caption" *ngIf="isObserving$" i18n>
```
Comment (Iloveall) [inline src/app/items/item-by-id.component.html:118]: Same here

File: src/app/items/item-by-id.component.html [hunk 5]
```diff
@@ -102,23 +102,23 @@
...
                 You are not allowed to view the progress of other users on this content.
               </p>
-              <p class="not-allow-caption" *ngIf="!isWatching" i18n>
+              <p class="not-allow-caption" *ngIf="!isObserving$" i18n>
```
Comment (Iloveall) [inline src/app/items/item-by-id.component.html:121]: Same here

**Review by Iloveall — CHANGES_REQUESTED** [review]
Huge PR 💪

**Comment by smadbe** [comment]
Thanks for spotting this error 👍 . Fixed

**Review by Iloveall — APPROVED** [review]

**Pull request merged**